### PR TITLE
feat: add json value converters

### PR DESCRIPTION
**Pull request by dongchangYoo**
We want to receive intermediate data for verifying the process of generating a proof as a JSON file. 
Therefore, in this PR, `RapidJsonValueConverter`s for the following four classes was implemented.
- PrimeFieldBase
- Extension Fields
- Elliptic Curve Points (only short weierstrass)
- Geometric Points
- UnivariatePolynomial (sparse, dense and their coefficients)
- UnivariateEvaluations

### CODE REVIEW

**Comment by chokobole** [comment]
You need to move all the `Copyable` implementations of points in d5e33cfc7557117cc23a0b16eda1fa8c653321a9

File: tachyon/math/polynomials/univariate/univariate_dense_polynomial_unittest.cc [hunk 1]
```diff
@@ -397,4 +397,19 @@ TEST_F(UnivariateDensePolynomialTest, Hash) {
...
 
+TEST_F(UnivariateDensePolynomialTest, JsonValueConverter) {
+  Poly expected_poly(Coeffs({GF7(1), GF7(2), GF7(3), GF7(4), GF7(5)}));
+  std::string expected_content =
+      R"({"coefficients":["0x1","0x2","0x3","0x4","0x5"]})";
+
+  Poly poly;
+  std::string error;
+  EXPECT_TRUE(base::ParseJson(expected_content, &poly, &error));
+  EXPECT_TRUE(error.empty());
```
Comment (chokobole) [inline tachyon/math/polynomials/univariate/univariate_dense_polynomial_unittest.cc:408]:
```suggestion
  ASSERT_TRUE(base::ParseJson(expected_content, &poly, &error));
  ASSERT_TRUE(error.empty());
```

File: tachyon/math/polynomials/univariate/univariate_polynomial.h [hunk 2]
```diff
@@ -294,6 +295,26 @@ class Copyable<math::UnivariatePolynomial<Coefficients>> {
...
+    AddJsonElement(object, "coefficients", value.coefficients().coefficients(), allocator);
+    return object;
+  }
+
+  static bool To(const rapidjson::Value& json_value, std::string_view key,
+                 math::UnivariatePolynomial<Coefficients>* value, std::string* error) {
+    using F = typename Coefficients::Field;
+    std::vector<F> coeffs;
+    if (!ParseJsonElement(json_value, "coefficients", &coeffs, error)) return false;
+    *value = math::UnivariatePolynomial(Coefficients(coeffs));
```
Comment (chokobole) [inline tachyon/math/polynomials/univariate/univariate_polynomial.h:319]:
```suggestion
    *value = math::UnivariatePolynomial(Coefficients(std::move(coeffs)));
```

**Comment by chokobole** [comment]
1a02e54ee0c78e9c77e8ea7ea84825dfae229200 only works when it's a `DenseUnivariatePolynomial`. Please write codes as we do with `Copyable`.

**Comment by chokobole** [comment]
Please add unittests in 1a02e54ee0c78e9c77e8ea7ea84825dfae229200 .

File: tachyon/math/polynomials/univariate/univariate_evaluations_unittest.cc [hunk 3]
```diff
@@ -275,4 +275,19 @@ TEST_F(UnivariateEvaluationsTest, Hash) {
...
 }
 
+TEST_F(UnivariateEvaluationsTest, JsonValueConverter) {
+  Poly expected_poly({GF7(1), GF7(2), GF7(3), GF7(4), GF7(5)});
+  std::string expected_content =
+      R"({"evaluations":["0x1","0x2","0x3","0x4","0x5"]})";
+
+  Poly poly;
+  std::string error;
+  EXPECT_TRUE(base::ParseJson(expected_content, &poly, &error));
```
Comment (chokobole) [inline tachyon/math/polynomials/univariate/univariate_evaluations_unittest.cc:285]:
```suggestion
  ASSERT_TRUE(base::ParseJson(expected_content, &poly, &error));
  ASSERT_TRUE(error.empty());
```

File: tachyon/math/polynomials/univariate/univariate_evaluations_unittest.cc [hunk 4]
```diff
@@ -275,4 +275,19 @@ TEST_F(UnivariateEvaluationsTest, Hash) {
                       Poly::Random(kMaxDegree))));
 }
 
+TEST_F(UnivariateEvaluationsTest, JsonValueConverter) {
+  Poly expected_poly({GF7(1), GF7(2), GF7(3), GF7(4), GF7(5)});
+  std::string expected_content =
```
Comment (chokobole) [inline tachyon/math/polynomials/univariate/univariate_evaluations_unittest.cc:280]:
```suggestion
  std::string expected_json =
```

**Comment by chokobole** [comment]
Some of changes in c8605d1485847ad9a3b4aa02f3936255da799de9 should be squashed.

File: tachyon/base/json/rapidjson_util.h [hunk 5]
```diff
@@ -177,7 +177,7 @@ class RapidJsonValueConverter<std::string> {
  public:
   template <typename Allocator>
   static rapidjson::Value From(const std::string& value, Allocator& allocator) {
-    return rapidjson::Value(value.c_str(), value.length());
+    return rapidjson::Value(value.c_str(), value.length(), allocator);
```
Comment (chokobole) [inline tachyon/base/json/rapidjson_util.h:180]: This should be separated into another commit.

**Comment by dongchangYoo** [comment]
> Please add unittests in [1a02e54](https://github.com/kroma-network/tachyon/commit/1a02e54ee0c78e9c77e8ea7ea84825dfae229200) .

The unittest is implemented at `univariate_dense_polynomial_unittest.cc`.

**Review by Insun35 — COMMENTED** [review]
Please change the scope of the commit message from `zk` to `math` in d5e33cf

File: tachyon/math/polynomials/univariate/univariate_polynomial.h [hunk 6]
```diff
@@ -294,6 +295,26 @@ class Copyable<math::UnivariatePolynomial<Coefficients>> {
...
+class RapidJsonValueConverter<math::UnivariatePolynomial<Coefficients>> {
+  public:
+    template <typename Allocator>
+  static rapidjson::Value From(const math::UnivariatePolynomial<Coefficients>& value, Allocator& allocator) {
+    rapidjson::Value object(rapidjson::kObjectType);
+    AddJsonElement(object, "coefficients", value.coefficients().coefficients(), allocator);
+    return object;
+  }
+
+  static bool To(const rapidjson::Value& json_value, std::string_view key,
```
Comment (Insun35) [inline tachyon/math/polynomials/univariate/univariate_polynomial.h:312]:
```suggestion
  [[nodiscard]] static bool To(const rapidjson::Value& json_value, std::string_view key,
```

File: tachyon/math/polynomials/univariate/univariate_evaluations.h [hunk 7]
```diff
@@ -252,6 +253,29 @@ class Copyable<math::UnivariateEvaluations<F, MaxDegree>> {
...
+  template <typename Allocator>
+  static rapidjson::Value From(
+      const math::UnivariateEvaluations<F, MaxDegree>& value,
+      Allocator& allocator) {
+    rapidjson::Value object(rapidjson::kObjectType);
+    AddJsonElement(object, "evaluations", value.evaluations(), allocator);
+    return object;
+  }
+
+  static bool To(const rapidjson::Value& json_value, std::string_view key,
```
Comment (Insun35) [inline tachyon/math/polynomials/univariate/univariate_evaluations.h:268]:
```suggestion
  [[nodiscard]] static bool To(const rapidjson::Value& json_value, std::string_view key,
```

File: tachyon/math/elliptic_curves/short_weierstrass/affine_point.h [hunk 8]
```diff
@@ -233,6 +234,33 @@ class Copyable<math::AffinePoint<
...
+  template <typename Allocator>
+  static rapidjson::Value From(const math::AffinePoint<Curve>& value,
+                               Allocator& allocator) {
+    rapidjson::Value object(rapidjson::kObjectType);
+    AddJsonElement(object, "x", value.x(), allocator);
+    AddJsonElement(object, "y", value.y(), allocator);
+    return object;
+  }
+
+  static bool To(const rapidjson::Value& json_value, std::string_view key,
```
Comment (Insun35) [inline tachyon/math/elliptic_curves/short_weierstrass/affine_point.h:253]:
```suggestion
  [[nodiscard]] static bool To(const rapidjson::Value& json_value, std::string_view key,
```

File: tachyon/math/polynomials/univariate/univariate_polynomial.h (hunk 6 above)
Comment (dongchangYoo) [reply tachyon/math/polynomials/univariate/univariate_polynomial.h:312]: Is it necessary to include the `nodiscard` keyword in utility functions? In my opinion, I'd prefer not to include it for user's convenience.

File: tachyon/math/polynomials/univariate/univariate_evaluations.h (hunk 7 above)
Comment (dongchangYoo) [reply tachyon/math/polynomials/univariate/univariate_evaluations.h:268]: Is it necessary to include the `nodiscard` keyword in utility functions? In my opinion, I'd prefer not to include it for user's convenience.

File: tachyon/math/elliptic_curves/short_weierstrass/affine_point.h (hunk 8 above)
Comment (dongchangYoo) [reply tachyon/math/elliptic_curves/short_weierstrass/affine_point.h:253]: Is it necessary to include the `nodiscard` keyword in utility functions? In my opinion, I'd prefer not to include it for user's convenience.

File: tachyon/math/polynomials/univariate/univariate_polynomial.h (hunk 6 above)
Comment (Insun35) [reply tachyon/math/polynomials/univariate/univariate_polynomial.h:318]: Make sense, cuz it's not as critical as core logic

**Comment by chokobole** [comment]
Please could you add unittests for `Fp3` and `Fp4` for c2c0d5778363bd1c32b57c213ee0a79b446b7012 ?

**Comment by dongchangYoo** [comment]
> Please could you add unittests for `Fp3` and `Fp4` for [c2c0d57](https://github.com/kroma-network/tachyon/commit/c2c0d5778363bd1c32b57c213ee0a79b446b7012) ?

Currently, There is no implementations of Fq3 and Fq4.

File: tachyon/base/json/rapidjson_util.h [hunk 9]
```diff
@@ -46,70 +47,76 @@ class RapidJsonValueConverter<bool> {
...
   static rapidjson::Value From(T value, Allocator& allocator) {
-    return rapidjson::Value(value);
+    static_assert(sizeof(T) <= sizeof(int));
+    return rapidjson::Value(static_cast<int>(value));
   }
 
   static bool To(const rapidjson::Value& json_value, std::string_view key,
                  T* value, std::string* error) {
-    if (!json_value.IsInt()) {
+    if (!json_value.IsUint()) {
```
Comment (chokobole) [inline tachyon/base/json/rapidjson_util.h:115]:
```suggestion
    if (!json_value.IsInt()) {
```

**Review by chokobole — APPROVED** [review]
LGTM

**Comment by TomTaehoonKim** [comment]
@dongchangYoo With [f070c8b](https://github.com/kroma-network/tachyon/pull/227/commits/f070c8b5c1d65d91e507b94adbbeeaa49933d70f), I have 2 comments
1. Is is a feat commit? I think it's a fix commit
2. a -> an

File: tachyon/base/json/rapidjson_util.h [hunk 10]
```diff
@@ -197,6 +197,25 @@ class RapidJsonValueConverter<std::string> {
...
+ public:
+  template <typename Allocator>
+  static rapidjson::Value From(std::string_view value, Allocator& allocator) {
+    return rapidjson::Value(value.data(), value.length());
+  }
+
+  static bool To(const rapidjson::Value& json_value, std::string_view key,
+                 std::string_view* value, std::string* error) {
+    if (!json_value.IsString()) {
+      *error = RapidJsonMismatchedTypeError(key, "string", json_value);
```
Comment (TomTaehoonKim) [inline tachyon/base/json/rapidjson_util.h:211]:
```suggestion
      *error = RapidJsonMismatchedTypeError(key, "string_view", json_value);
```

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

**Comment by dongchangYoo** [comment]
> @dongchangYoo With [f070c8b](https://github.com/kroma-network/tachyon/pull/227/commits/f070c8b5c1d65d91e507b94adbbeeaa49933d70f), I have 2 comments
> 
> 1. Is is a feat commit? I think it's a fix commit
> 2. a -> an

The implementation has been modified to copy the input string by adding an `allocator` as an input. This change in handling strings, I believe, qualifies as a 'feat'.

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM
